File: Practice/Python/Data Structure/Tree-Based/TreeTraversal.py
# Tree traversal in Python (DFS) - Inorder, Preorder, Postorder
# Tree traversal is a form of graph traversal. It is the process of visiting each node in a tree exactly once. Unlike linear data structures (Array, Linked List, Queues, Stacks, etc) which have only one logical way to traverse them, trees can be traversed in different ways. Following are the generally used ways for traversing trees.
class Node:
    def __init__(self, item):
        self.left = None
        self.right = None
        self.val = item
def inorder(root):
    if root:
        inorder(root.left) # Traverse left
        print(str(root.val) + "->", end='') # Traverse root
        inorder(root.right) # Traverse right
def postorder(root):
    if root:
        postorder(root.left) # Traverse left
        postorder(root.right) # Traverse right
        print(str(root.val) + "->", end='')# Traverse root
def preorder(root):
    if root:
        print(str(root.val) + "->", end='') # Traverse root
        preorder(root.left) # Traverse left
        preorder(root.right) # Traverse right

root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.right = Node(5)

print("Inorder traversal ")
inorder(root)
print("\nPreorder traversal ")
preorder(root)
print("\nPostorder traversal ")
postorder(root)

File: Practice/Java/Data Structure/Tree-Based/TreeTraversal.java
// Tree traversal is a form of graph traversal. It is the process of visiting each node in a tree exactly once. Unlike linear data structures (Array, Linked List, Queues, Stacks, etc) which have only one logical way to traverse them, trees can be traversed in different ways. Following are the generally used ways for traversing trees.

/**
 * In-order Traversal (Left-Root-Right)
 * 
 * in-order(root->left)
 * display(root->data)
 * in-order(root->right)
 * 
 * Pre-order Traversal (Root-Left-Right)
 * 
 * display(root->data)
 * pre-order(root->left)
 * pre-order(root->right)
 * 
 * Post-order Traversal (Left-Right-Root)
 * 
 * post-order(root->left)
 * post-order(root->right)
 * display(root->data)
*/

class Node {
    int item;
    Node left, right;

    public Node(int key) {
        item = key;
        left = right = null;
    }
}

class BinaryTree {
    // Root of Binary Tree
    Node root;

    BinaryTree() {
        root = null;
    }

    void postorder(Node node) {
        if (node == null)
            return;

        // Traverse left
        postorder(node.left);
        // Traverse right
        postorder(node.right);
        // Traverse root
        System.out.print(node.item + "->");
    }

    void inorder(Node node) {
        if (node == null)
            return;

        // Traverse left
        inorder(node.left);
        // Traverse root
        System.out.print(node.item + "->");
        // Traverse right
        inorder(node.right);
    }

    void preorder(Node node) {
        if (node == null)
            return;

        // Traverse root
        System.out.print(node.item + "->");
        // Traverse left
        preorder(node.left);
        // Traverse right
        preorder(node.right);
    }

    public static void main(String[] args) {
        BinaryTree tree = new BinaryTree();
        tree.root = new Node(1);
        tree.root.left = new Node(12);
        tree.root.right = new Node(9);
        tree.root.left.left = new Node(5);
        tree.root.left.right = new Node(6);

        System.out.println("Inorder traversal");
        tree.inorder(tree.root);

        System.out.println("\nPreorder traversal ");
        tree.preorder(tree.root);

        System.out.println("\nPostorder traversal");
        tree.postorder(tree.root);
    }
}